kyo-core/shared/src/main/scala/kyo/Maybe.scala

```diff
@@ -0,0 +1,159 @@
+package kyo
+
+import Maybe.*
+import Maybe.internal.*
+
+opaque type Maybe[+T] >: (Empty | Defined[T]) = Empty | Defined[T]
+
+object Maybe:
+    given [T, U](using CanEqual[T, U]): CanEqual[Maybe[T], Maybe[U]] = CanEqual.derived
+    given [T]: Conversion[Maybe[T], IterableOnce[T]]                 = _.iterator
+    given [T: Flat]: Flat[Maybe[T]]                                  = Flat.unsafe.bypass
+
+    def apply[T](v: T): Maybe[T] =
+        if isNull(v) then
+            Empty
+        else
+            v match
+                case v: DefinedEmpty => v.nest
+                case v               => v
+
+    def fromScala[T](opt: Option[T]): Maybe[T] =
+        opt match
+            case Some(v) => Defined(v)
+            case None    => Empty
+
+    def empty[T]: Maybe[T] = Empty
+
+    def when[T](cond: Boolean)(v: => T): Maybe[T] =
+        if cond then v else Empty
+
+    opaque type Defined[+T] = T | DefinedEmpty
+
+    object Defined:
+
+        def apply[T](v: T): Defined[T] =
+            v match
+                case v: DefinedEmpty => v.nest
+                case v: Empty        => DefinedEmpty()
+                case v               => v
+
+        // TODO avoid allocation
+        def unapply[T](opt: Maybe[T]): Option[T] =
+            opt.toScala
+    end Defined
+
+    sealed abstract class Empty
+    case object Empty extends Empty
+
+    extension [T](self: Maybe[T])
+
+        inline def toScala: Option[T] =
+            if isEmpty then None
+            else Some(get)
```

Comment:
Just noting that this is only shallow unpacking... it may be annoying in the case of non-trivial nesting. Though I suspect there is rarely a usecase where you want to unpack many nested Maybes to nested Options. The conversion would be quite expensive.

Reply:
I think we'd need a macro to transform nested types

Reply:
Couldn't it be recursive?

Reply:
The issue is that we'd need to change the type of the nested values and the method operates only at "one level". `Maybe(Maybe(1)).toOption == Option(Maybe(1))`. It becomes even more complicated if it's in a generic scope `def test[T](v: Maybe[T]): Option[T] = v.toOption; test(Maybe(Maybe(1))`